### ISSUE TITLE
Błędy Access Denied w /shipment-management/shipments/create-commands

### ISSUE THREAD

**Issue by tztztztz** [allegro/allegro-api#9793]
Występuje problem: 
 - [x] z zasobem REST API / with a REST API resource
 - [x] ze środowiskiem testowym / with the test environment

:beetle: **Opis / Description**

Dzień dobry,

na TESTOWYM ALLEGRO wysyłam na endpoint /shipment-management/shipments/create-commands metodą POST dane do stworzenia paczki.

Otrzymuję błędy:

{"errors":[{"code":"AccessDenied","message":"Access is denied","details":null,"path":null,"userMessage":"No access to the specified resource.","metadata":{}}]}

Na pewno jest wszystko dobrze z autoryzacją / tokenem, ponieważ jeżeli za moment na tym samym tokenie tworzę paczkę starym sposobem ( PUT /parcel-management/parcel-create-commands/{commandId} ),

to nie ma żadnych problemów.

trace-id to ecd7d82851e1b5de

Proszę o sprawdzenie co może być przyczyną tego błędu.

**Comment by PrzemyslawLukanowski** [comment]
Dzień dobry, nowe zasoby shipment-management wymagają zakresów `allegro:api:shipments:write` i `allegro:api:shipments:read`. Uzupełnij je na stronie https://apps.developer.allegro.pl.allegrosandbox.pl/ i następnie wygeneruj nową parę tokenów poprzez ponowną autoryzację aplikacji przez użytkownika. Nowo wygenerowane tokeny będą posiadać już uwzględniony nowe zakres i tym samym będziesz mógł korzystać ze wskazanego zasobu.

**Comment by tztztztz** [comment]
Rozumiem, że to samo muszę zrobić później na wersji produkcyjnej?

**Comment by PrzemyslawLukanowski** [comment]
Dokładnie tak.

**Comment by tztztztz** [comment]
Już chciałem zamknąć tan Issue, ale teraz kiedy wysyłam dane paczki dostaję taki błąd:

{"commandId":"01c15618-e32c-4192-aea5-3dc99edd0ea4","status":"ERROR","errors":[{"code":"CARRIER_API_CLIENT_UNAVAILABLE","message":"Carrier Api Client is temporary unavailable - try again later or report problem","details":"Carrier Api Client is temporary unavailable - try again later or report problem","path":"","userMessage":"API przewoźnika jest tymczasowo niedostępne. Ponów próbę później."}],"shipmentId":null}

"API przewoźnika jest tymczasowo niedostępne. Ponów próbę później"

Rozumiem, że ja nic nie zawiniłem, tylko to jest coś po waszej stronie? 

Kiedy można się spodziewać naprawienia problemu?

**Comment by PrzemyslawLukanowski** [comment]
Tak, wygląda to na problem z testowym API DHL. Weryfikuję ten temat, wrócę z informacją.

**Comment by tztztztz** [comment]
Niestety, nadal to samo:

{"commandId":"9c317525-227f-44b2-9442-eda184cafb07","status":"ERROR","errors":[{"code":"CARRIER_API_CLIENT_UNAVAILABLE","message":"Carrier Api Client is temporary unavailable - try again later or report problem","details":"Carrier Api Client is temporary unavailable - try again later or report problem","path":"","userMessage":"API przewoźnika jest tymczasowo niedostępne. Ponów próbę później."}],"shipmentId":null}

trace-id c3ae2bd11d556ef9

**Comment by PrzemyslawLukanowski** [comment]
Poczekaj jeszcze proszę ok. 15 minut i ponów próbę.

**Comment by PrzemyslawLukanowski** [comment]
Zespół potwierdził, że problem jednak powrócił. Przepraszam za niedogodności, ponownie wrócę z informacją, gdy problem zostanie rozwiązany.

**Comment by tztztztz** [comment]
Rozumiem, czekam zatem na informację.

**Comment by PrzemyslawLukanowski** [comment]
Zostały wprowadzony pewne poprawki, jednak otrzymałem informację, że mimo wszystko w niektórych przypadkach dziś wciąż mogą występować okresowe problemy po stronie API DHL.

**Comment by tztztztz** [comment]
Paczkę udaje się stworzyć, ale kiedy próbuję anulować paczkę, zostaje zwracany błąd: "SHIPMENT_NOT_FOUND"

 [Array ( [0] => stdClass Object ( [code] => SHIPMENT_NOT_FOUND [message] => Requested shipment was not found [details] => Requested shipment was not found [path] => shipmentId [userMessage] => Nie odnaleziono przewozu ) ) ]
]

czy jest szansa, że jutro wszystkie funkcje będą działać prawidłowo?

**Comment by PrzemyslawLukanowski** [comment]
Czy możesz podać trace-id lub command-id tej operacji. To niekoniecznie musi być związane z problemami po stronie DHL, zweryfikuję to.

**Comment by tztztztz** [comment]
Wykonałem operację jeszcze raz, i przy próbie weryfikacji anulowania występuje powyższy błąd.

trace-id
ecf05eee6ed876ea

**Comment by PrzemyslawLukanowski** [comment]
W tym przypadku komunikat jest poprawny, ponieważ wartość shipmentId nie jest prawidłowa ("90029345580" to waybill). ShipmentId to UUID, który otrzymujesz w odpowiedzi [GET /shipment-management/shipments/create-commands/{commandId}](https://developer.allegro.pl/documentation#operation/getShipmentCreationStatus):

<img width="524" alt="Zrzut ekranu 2024-09-2 o 13 22 32" src="https://github.com/user-attachments/assets/94fd36ba-54f4-4b54-9e4a-05cdee882799">

**Comment by tztztztz** [comment]
W takim razie po stworzeniu paczki, powinienem dodatkowo uruchomić metodę 

GET /shipment-management/shipments/create-commands/{commandId}

i zapisać shipmentId w celu anulowania przesyłki lub np. pobrania etykiety?

**Comment by tztztztz** [comment]
Tylko pytanie jak jest tutaj z synchronizacją, ponieważ widzę, że w tej metodzie GET nie ma identyfikatora listu przewozowego, a jednocześnie / asynchronicznie może być tworzonych wiele przesyłek, jak mam teraz skojarzyć shipmentId z określonym waybill?

**Comment by PrzemyslawLukanowski** [comment]
> GET /shipment-management/shipments/create-commands/{commandId}
> 
> i zapisać shipmentId w celu anulowania przesyłki lub np. pobrania etykiety?

Dokładnie tak, lub np. w celu sprawdzenia szczegółów utworzonej przesyłki.

> Tylko pytanie jak jest tutaj z synchronizacją, ponieważ widzę, że w tej metodzie GET nie ma identyfikatora listu przewozowego, a jednocześnie / asynchronicznie może być tworzonych wiele przesyłek, jak mam teraz skojarzyć shipmentId z określonym waybill?

Po otrzymaniu statusu SUCCESS w GET /shipment-management/shipments/create-commands/{commandId} możesz odpytać o szczegóły danego shipmentu przez [GET /shipment-management/shipments/{shipmentId}](https://developer.allegro.pl/documentation/#operation/getShipmentDetails) - w odpowiedzi zwrócimy waybill w polu `packages.waybill`. W ten sposób możesz właśnie skojarzyć dany shipmentId z waybillem.

**Comment by tztztztz** [comment]
Niestety pojawia się kolejny problem. 

Otóż powiedzmy, że w związku z tym jak jest to zrobione, nie będę od razu, w tym samym wątku serwera www, po zrobieniu paczki wywoływał GET /shipment-management/shipments/create-commands/{commandId}, bo nie wiadomo jak długo może to trwać, a może trwać tak długo, że nastąpi rozłącznie połączenia HTTP. 

Będzie za to odpowiedzialny osobny proces, który w tle, już nie w trybie HTTP, będzie wywoływał metodę GET /shipment-management/shipments/create-commands/{commandId}

I teraz pytanie 1) jak to właściwie będzie działać - czy po tym jak wywołam /parcel-management/parcel-create-commands/{commandId} i dostanę status SUCCESS, to przy kolejnych wywołaniach tej metody już nigdy więcej nie dostanę informacji o tej samej paczce, i powinienem tak długo wywoływać metodę aż zwróci... no właśnie co - pusty wynik, czy jakiś określony kod odpowiedzi, oznaczający, że nie ma już więcej paczek do przetworzenia.

Pytanie 2) rozumiem że "parcelId" w odpowiedzi z GET /shipment-management/shipments/create-commands/{commandId}, to jest "shipmentId" który mam później używać dla danego waybill?

Sugestia: całość byłaby znacznie prostsza gdybym mógł odpytać o status utworzenia paczki wg. waybill, a nie shipmentId, które dostaję dopiero po potwierdzeniu stworzenia paczki.

**Comment by PrzemyslawLukanowski** [comment]
CommandID jest jednokrotnego użytku, tzn. że dla każdej nowej przesyłki, którą będziesz generować przez POST /shipment-management/shipments/create-commands będziemy zwracać nowe commandId. Jeśli przekazujesz nam swoje commandId w body (w tym zasobie nie musisz tego robić. Jeśli go nie przekażesz, sami je wygenerujemy w odpowiedzi), to dla nowej przesyłki musisz wygenerować nowy UUID. Dlatego:

> dostanę status SUCCESS, to przy kolejnych wywołaniach tej metody już nigdy więcej nie dostanę informacji o tej samej paczce, i powinienem tak długo wywoływać metodę aż zwróci... no właśnie co - pusty wynik, czy jakiś określony kod odpowiedzi, oznaczający, że nie ma już więcej paczek do przetworzenia? Muszę wiedzieć kiedy już dalej nie wywoływać tej metody.

dla danego commandId zawsze będziemy zwracać info o tej samej przesyłce. Status SUCCESS będzie świadczyć o tym, że nie musisz już odpytywać danego commandId, bo kolejne będą zwracać tę samą informację co wcześniej.

**Comment by tztztztz** [comment]
Ok, nie byłem pewien czy "commandId" zwracany przy tworzeniu paczki można potem użyć do sprawdzenia statusu, w takim razie to upraszcza sprawę.

**Comment by tztztztz** [comment]
Mam jeszcze problem z zamówieniem z punktem odbioru (pickup).

Przy próbie weryfikacji stworzenia paczki dostaję błąd "Nie znaleziono punktu odbioru (id: Błąd - nieznany numer SAP)"

```json
{"commandId":"a8c96504-dd81-4434-89cd-ce93398551ce","status":"ERROR","errors":[{"code":"POINT_NOT_FOUND","message":"Pick-up point (id: Błąd - nieznany numer SAP) not found","details":"","path":"","userMessage":"Nie znaleziono punktu odbioru (id: Błąd - nieznany numer SAP)"}],"shipmentId":null}
```

trace-id 9dea393944c98fdc

Proszę o sprawdzenie co jest nie tak.

**Comment by PrzemyslawLukanowski** [comment]
Na Sandboxie nie wszystkie punkty są dostępne, spróbuj skorzystać z 4591097 lub 4509039.

**Comment by tztztztz** [comment]
Podczas tworzenia paczki występuje błąd:

"userMessage": "Adres i numer telefonu dla podjazdu musi by\u0107 taki sam jak nadawcy"

Z tego wynika, że input->pickup wysyłane w danych paczki to powinna być kopia input->sender, a dane w "checkout form" zawarte w delivery->pickupPoint z których tworzymy paczkę, powinny zostać zignorowane?

Jedynie wypełnienie pola "point" w input->pickup (oraz input->receiver) tworzyłoby różnicę między "sender" a "pickup"?

**Comment by PrzemyslawLukanowski** [comment]
> Z tego wynika, że input->pickup wysyłane w danych paczki to powinna być kopia input->sender,

Dokładnie tak.

> a dane w "checkout form" zawarte w delivery->pickupPoint z których tworzymy paczkę, powinny zostać zignorowane?

> Z delivery->pickupPoint natomiast wyciągamy jedynie dane dla pola "point" w input->pickup (oraz input->receiver) i jedynie to tworzyłoby różnicę między danymi w "sender" a "pickup"?


Tu dopowiem, że input.pickup to adres, z którego przesyłka ma zostać odebrana od nadawcy. To pole nie jest tożsame z delivery.pickupPoint z zamówienia - ta informacja dotyczy punktu odbiorczego, do którego ma trafić przesyłka.

**Comment by tztztztz** [comment]
Rozumiem, ale w takim razie dane z pickupPoint nigdzie nie pojawiają się w danych paczki do zrobienia?

Używamy jedynie delivery->pickupPoint->id dla pickup->point i receiver->point, i to wystarczy do określenia punktu odbioru dla klienta?

**Comment by PrzemyslawLukanowski** [comment]
> Rozumiem, ale w takim razie dane z pickupPoint nigdzie nie pojawiają się w danych paczki do zrobienia?

Nie, ponieważ, to sprzedawca decyduje, w jakim punkcie nada przesyłkę

> Używamy jedynie delivery->pickupPoint->id dla pickup->point i receiver->point, i to wystarczy do określenia punktu odbioru dla klienta?

Dla pickup->point nie, ponieważ jest to punk nadania, po stronie sprzedawcy, nie kupującego. Dla receiver->point jak najbardziej, ponieważ wskazuje to punkt odbioru.

**Comment by tztztztz** [comment]
Ok, tylko w dokumentacji pisze coś takiego przy polu pickup->point:

"Pickup point id. You can get it from order or directly from courier service."

to wskazuje jakby chodziło o wyciągnięcie i skopiowanie "delivery->pickupPoint->id" nie tylko do "receiver->point" ale także do "pickup->point".

Rozumiem jednak, że nie muszę wypełniać tego pola, i wystarczy receiver->point aby określić np. paczkomat inpost który klient sobie wybrał i który jest wpisany w "delivery->pickupPoint->id" w checkout form zamówienia klienta?

**Comment by PrzemyslawLukanowski** [comment]
Dziękuję za zwrócenie na to uwagi, rzeczywiście mogło to wprowadzić w błąd. Będziemy poprawiać ten opis.

Tak, nie musisz, jeśli nie będziesz/sprzedawca nie będzie nadawać przesyłki w żadnym punkcie, tylko np. kurier podjedzie pod konkretny adres. To już w zależności od tego, gdzie następuje odbiór przesyłek do nadania. Poniżej opis z naszego [poradnika](https://developer.allegro.pl/tutorials/jak-zarzadzac-przesylkami-przez-wysylam-z-allegro-LRVjK7K21sY#jak-utworzyc-nowa-paczke):

```
      "pickup":{ - niewymagane, dane miejsca przekazania przesyłki
         "name":"Jan Kowalski", - dane osobowe nadawcy
         "company":"Allegro.pl sp. z o.o.", - nazwa firmy
         "street":"Główna", - ulica
         "streetNumber":"30", - numer budynku
         "postalCode":"10-200", - kod pocztowy
         "city":"Warszawa", - miasto
         "countryCode":"PL", - kod kraju zgodny ze standardem ISO 3166-1 alpha-2
         "email":"email@mail.com", - adres e-mail
         "phone":"500600700", - numer telefonu nadawcy
         "point":"A1234567" - wymagane, jeśli adresem nadawczym jest punkt odbioru
      },
```

Całe pole `pickup` oznaczone jest jako niewymagane, ponieważ jeśli go nie wypełnisz, użyjemy danych z `sender`.

**Comment by tztztztz** [comment]
Ok, dzięki za info, w takim razie w ogóle nie będę wypełniał pickup, a w przypadku, kiedy klient wybierze sobie jakiś punkt odbioru typu Inpost, id tego punktu będzie w receiver->point.
